### ISSUE TITLE
Inconsistent Results in SELECT

### ISSUE THREAD

**Issue by bajinsheng** [pingcap/tidb#36853]
## Bug Report

There is an inconsistent result returned by SELECT statements. 

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE t0(c0 INT);
CREATE TABLE t1(c0 INT);
CREATE VIEW v0(c0) AS SELECT 1 FROM t0, t1 WHERE 1;
INSERT INTO t0 VALUES (2);
REPLACE INTO t1 VALUES ('2');

SELECT t0.c0 FROM t0, v0; -- {2}

SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false); -- {2}
SELECT t0.c0 FROM t0, v0 WHERE NOT (ELT('1', v0.c0>false)); -- {2}
SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false) IS NULL; -- {2}
```

### 2. What did you expect to see? (Required)

The last three SELECTs return more results than expected. These conditions in the WHERE clause are mutually exclusive, but they all return the same results.

### 3. What did you see instead (Required)
The last three SELECTs all return {2}.

### 4. What is your TiDB version? (Required)

| Release Version: v6.2.0-alpha-584-gb1edc33fc
Edition: Community
Git Commit Hash: b1edc33fc1a99f69d0527478ddb082d7592af74e
Git Branch: master
UTC Build Time: 2022-08-02 08:22:35
GoVersion: go1.18.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore |

**Comment by tiancaiamao** [comment]
The first query `SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false); -- {2}` works as expected.

Test in MySQL:

```
mysql> SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false);
+------+
| c0   |
+------+
|    2 |
+------+
1 row in set (0.00 sec)

mysql> explain SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false);
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                         |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------------------------+
|  1 | SIMPLE      | t0    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | NULL                          |
|  1 | SIMPLE      | t0    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using join buffer (hash join) |
|  1 | SIMPLE      | t1    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using join buffer (hash join) |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------------------------+
3 rows in set, 1 warning (0.00 sec)
```

In TiDB:

```
mysql> SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false);
+------+
| c0   |
+------+
|    2 |
+------+
1 row in set (0.00 sec)

mysql> explain SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false);
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| id                                | estRows | task      | access object | operator info                  |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| HashJoin_14                       | 1.00    | root      |               | CARTESIAN inner join           |
| ├─Projection_18(Build)            | 1.00    | root      |               | 1->Column#9                    |
| │ └─HashJoin_19                   | 1.00    | root      |               | CARTESIAN inner join           |
| │   ├─TableReader_24(Build)       | 1.00    | root      |               | data:TableFullScan_23          |
| │   │ └─TableFullScan_23          | 1.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo |
| │   └─TableReader_22(Probe)       | 1.00    | root      |               | data:TableFullScan_21          |
| │     └─TableFullScan_21          | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
| └─TableReader_17(Probe)           | 1.00    | root      |               | data:TableFullScan_16          |
|   └─TableFullScan_16              | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
9 rows in set (0.00 sec)
```

**Comment by tiancaiamao** [comment]
The second and third are wrong.

MySQL:

```
mysql> SELECT t0.c0 FROM t0, v0 WHERE NOT (ELT('1', v0.c0>false));
Empty set (0.00 sec)
mysql> explain SELECT t0.c0 FROM t0, v0 WHERE NOT (ELT('1', v0.c0>false));
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra            |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
|  1 | SIMPLE      | NULL  | NULL       | NULL | NULL          | NULL | NULL    | NULL | NULL |     NULL | Impossible WHERE |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
1 row in set, 1 warning (0.01 sec)

mysql> SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false) IS NULL;
Empty set (0.00 sec)
mysql> explain SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false) IS NULL;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra            |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
|  1 | SIMPLE      | NULL  | NULL       | NULL | NULL          | NULL | NULL    | NULL | NULL |     NULL | Impossible WHERE |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+------------------+
1 row in set, 1 warning (0.00 sec)
```

But in TiDB:

```
mysql> SELECT t0.c0 FROM t0, v0 WHERE NOT (ELT('1', v0.c0>false));
+------+
| c0   |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
mysql> explain SELECT t0.c0 FROM t0, v0 WHERE NOT (ELT('1', v0.c0>false));
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| id                                | estRows | task      | access object | operator info                  |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| HashJoin_14                       | 1.00    | root      |               | CARTESIAN inner join           |
| ├─Projection_18(Build)            | 1.00    | root      |               | 1->Column#9                    |
| │ └─HashJoin_19                   | 1.00    | root      |               | CARTESIAN inner join           |
| │   ├─TableReader_24(Build)       | 1.00    | root      |               | data:TableFullScan_23          |
| │   │ └─TableFullScan_23          | 1.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo |
| │   └─TableReader_22(Probe)       | 1.00    | root      |               | data:TableFullScan_21          |
| │     └─TableFullScan_21          | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
| └─TableReader_17(Probe)           | 1.00    | root      |               | data:TableFullScan_16          |
|   └─TableFullScan_16              | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
9 rows in set (0.00 sec)

mysql> SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false) IS NULL;
+------+
| c0   |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
mysql> explain SELECT t0.c0 FROM t0, v0 WHERE ELT('1', v0.c0>false) IS NULL;
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| id                                | estRows | task      | access object | operator info                  |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
| HashJoin_14                       | 1.00    | root      |               | CARTESIAN inner join           |
| ├─Projection_18(Build)            | 1.00    | root      |               | 1->Column#9                    |
| │ └─HashJoin_19                   | 1.00    | root      |               | CARTESIAN inner join           |
| │   ├─TableReader_24(Build)       | 1.00    | root      |               | data:TableFullScan_23          |
| │   │ └─TableFullScan_23          | 1.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo |
| │   └─TableReader_22(Probe)       | 1.00    | root      |               | data:TableFullScan_21          |
| │     └─TableFullScan_21          | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
| └─TableReader_17(Probe)           | 1.00    | root      |               | data:TableFullScan_16          |
|   └─TableFullScan_16              | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
+-----------------------------------+---------+-----------+---------------+--------------------------------+
9 rows in set (0.00 sec)
```

**Comment by qw4990** [comment]
A minimal reproducible case:
```
SELECT * FROM  v0 WHERE ELT('1', v0.c0>false);
SELECT * FROM  v0 WHERE NOT ELT('1', v0.c0>false); 
```

It's caused by the join reorder. As you can see below, after reordering, the condition is lost:
```
before join reorder >>  *core.LogicalProjection
before join reorder >>    *core.LogicalProjection
before join reorder >>       Join  [] [] [] [not(cast(elt(1, cast(gt(Column#5, 0), var_string(20))), double BINARY))]
before join reorder >>         DS  []
before join reorder >>         DS  []

after join reorder >>  *core.LogicalProjection
after join reorder >>    *core.LogicalProjection
after join reorder >>       Join  [] [] [] []
after join reorder >>         DS  []
after join reorder >>         DS  []
```

**Comment by winoros** [comment]
This issue is fixed by #37117.
It's caused by that the column substitute is not full successfully executed. But we think it is, and we used the replaced result.
And the replaced result is nil since it's not full successfully executed. We got the wrong result finally. That pr is to solve this bug.

**Comment by winoros** [comment]
I update the title, and wish that the scenario can be clearer.
